WebSockets: refactor for error handling, no global mutable variables, etc

adnelson opened the pull request:
This is a refactor of the websockets subscription module.

Changes:

* Remove global mutable variables and unsafePerformIO, which follows best practices and allows for more than one websocket in an app
* Enable error handling, allowing the user to e.g. respond to a situation where the websocket fails to connect, or a message fails to parse
* Allow the reconnection behavior to be configurable by having the user pass in a function which can decide whether to reconnect
* Enable a simple ping behavior which can be useful to prevent the browser from closing an inactive socket
* Prevent invalid behavior such as closing a websocket before it has opened or sending a message to a closed socket, by using MVar. Also there’s no need for a thread to poll whether the socket has closed
* Allow apps to send plain text as well as ToJSON objects

Despite the large list above, the external interface has only changed a bit. As can be seen in the updated websockets example, there are relatively few things that need to be changed in current code. The main change is that the user is responsible for creating the socket object, which must also be used when sending data.

cocreature commented on ghcjs-src/Miso/FFI.hs:
```diff
@@ -101,23 +104,41 @@ foreign import javascript unsafe "console.log($1);"
+-- | than 'parseString'.
+parseString :: FromJSON json => JSString -> IO json
+{-# INLINE parseString #-}
+parseString = parse . unsafeCoerce
```
No need for `unsafeCoerce`, you can use `jsval` here.

adnelson replied:
Didn't know about that one, thanks!

cocreature commented on ghcjs-src/Miso/FFI.hs:
```diff
@@ -101,23 +104,41 @@ foreign import javascript unsafe "console.log($1);"
 
-foreign import javascript unsafe "$r = JSON.parse($1);"
+-- | Parse a raw string into JSON. Accepts any JS value.
+foreign import javascript unsafe "$r = JSON.parse('' + $1);"
```
This seems fishy, when is this needed?

adnelson replied:
So the problem there is that `JSON.parse()` expects a string, but the type signature for `parse'` accepts any `JSVal`. So you can get obscure exceptions thrown.

Admittedly the better solution would be to refactor the code so that `parse'` takes ` JSString` instead, and/or to wrap the `JSON.parse()` in a `try`/`catch` so that we can handle a JSON parse exception. I was actually planning on doing that but didn't get there because I didn't want this PR to get too large. I could certainly revert this change, but on the other hand, it doesn't make the code any *less* safe than it already is.

cocreature commented on ghcjs-src/Miso/FFI.hs:
```diff
@@ -101,23 +104,41 @@ foreign import javascript unsafe "console.log($1);"
+-- | Parse a JSString into a result, so that it can be checked.
+safeParseString :: FromJSON obj => JSString -> IO (Result obj)
+{-# INLINE safeParseString #-}
+safeParseString = safeParse . unsafeCoerce
```
Same here.

cocreature commented on ghcjs-src/Miso/FFI.hs:
```diff
@@ -101,23 +104,41 @@ foreign import javascript unsafe "console.log($1);"
+parseString = parse . unsafeCoerce
+
+-- | Parses a JSVal (which of course should be a string). Less safe
+-- | than 'parseString'.
```
This should reference to `safeParseString`. I would also suggest to remove the `safe` prefix and instead add an `unsafe` prefix to this function. The docs should also mention in what way it is unsafe.

adnelson replied:
Totally in favor of that; I just didn't want to rename a function which is used throughout the code. Can do though.

cocreature commented on ghcjs-src/Miso/Subscription/WebSocket.hs:
```diff
@@ -151,48 +129,207 @@ data SocketState
-  Just ws <- readIORef websocket
-  toEnum <$> getSocketState' ws
+-- | Low-level websocket handle, wraps the raw JSVal.
+newtype WebSocket_ = WebSocket_ JSVal
```
Maybe call this `JSWebSocket` to make it easier to infer what this represents based on the name.

adnelson replied:
Sounds good

cocreature commented on ghcjs-src/Miso/Subscription/WebSocket.hs:
```diff
@@ -151,48 +129,207 @@ data SocketState
+  let (url, protocols) = (wsURL socket, wsProtocols socket)
+
+  -- Spin off ping thread if configured
+  pingThreadId <- case websocketPing config of
```
You can simplify this using `for`.

adnelson replied:
simpler, but less readable? I guess I'm fine with it but sometimes using combinators obscures things.

cocreature commented on ghcjs-src/Miso/Subscription/WebSocket.hs:
```diff
@@ -151,48 +129,207 @@ data SocketState
+  onClose socket_ =<< do
+    asyncCallback1 $ \e -> do
+      -- Stop the ping thread, if it's running.
+      maybe (pure ()) killThread pingThreadId
```
`traverse_ killThread pingThreadId`

adnelson replied:
works for me

adnelson replied:
Actually I'd kind of prefer using `mapM_` since it's part of the prelude

cocreature commented on ghcjs-src/Miso/Subscription/WebSocket.hs:
```diff
@@ -151,48 +129,207 @@ data SocketState
+
+-- | Create a new low-level websocket.
+newWebSocket_ :: MisoString -> [MisoString] -> IO (Either MisoString WebSocket_)
+{-# INLINE newWebSocket_ #-}
```
I’m a bit suspicious of all these `INLINE` pragmas. They tend to be a good way to ruin compile times.

adnelson replied:
The original had these on many of the functions so I was just following the convention. I'm totally fine with removing these.

adnelson commented:
@cocreature thanks for taking a look at this!

* I renamed the JSON parse functions, and made the ones that I had written safer (by catching `JSON.parse()` failures). This added some SLOC and possibly would be better in a separate PR, but I just figured I'd add it here.
* Renamed `WebSocket_` -> `JSWebSocket`
* Use `mapM` and `mapM_` (if you feel strongly about `for` or `traverse_`, I can use them)
* Since the lower-level parse functions now take `JSString` instead of `JSVal`, I still have to use `unsafeCoerce` :( but, sooner or later those functions might be able to be gotten rid of anyway and only the safe `JSString` functions used.
* Removed some of the `INLINE` pragmas but kept others. I really don't understand the tradeoffs well enough, open to inlining whatever, or not.
